test/integration/joint_state_publisher_system.cc

```diff
@@ -0,0 +1,132 @@
+class JointStatePublisherTest : public ::testing::TestWithParam<int>
+{
+  // Documentation inherited
+  protected: void SetUp() override
+  {
+    common::Console::SetVerbosity(4);
+    setenv("IGN_GAZEBO_SYSTEM_PLUGIN_PATH",
+           (std::string(PROJECT_BINARY_PATH) + "/lib").c_str(), 1);
+  }
+};
+
+/////////////////////////////////////////////////
+TEST_F(JointStatePublisherTest, DefaultPublisher)
+{
+  // Start server
+  ServerConfig serverConfig;
+  serverConfig.SetSdfFile(std::string(PROJECT_SOURCE_PATH) +
+      "/test/worlds/diff_drive.sdf");
+
+  Server server(serverConfig);
+  EXPECT_FALSE(server.Running());
+  EXPECT_FALSE(*server.Running(0));
+
+  server.SetUpdatePeriod(0ns);
+
+  // Check that all of joints are published.
+  std::function<void(const msgs::Model &)> jointStateCb =
+    [&](const msgs::Model &_msg)
+    {
+      bool foundLeftWheelJoint{false},
+           foundRightWheelJoint{false},
+           foundCasterWheel{false},
+           extra{false};
+
+      for (int i = 0; i < _msg.joint_size(); ++i)
+      {
+        if (_msg.joint(i).name() == "left_wheel_joint")
+          foundLeftWheelJoint = true;
+        else if (_msg.joint(i).name() == "right_wheel_joint")
+          foundRightWheelJoint = true;
+        else if (_msg.joint(i).name() == "caster_wheel")
+          foundCasterWheel = true;
+        else
+          extra = true;
+      }
+      EXPECT_TRUE(foundLeftWheelJoint);
+      EXPECT_TRUE(foundRightWheelJoint);
+      EXPECT_TRUE(foundCasterWheel);
+      EXPECT_FALSE(extra);
+    };
+
+  transport::Node node;
+  node.Subscribe("/world/diff_drive/model/vehicle/joint_state", jointStateCb);
+
+  server.Run(true, 10, false);
+}
+
+/////////////////////////////////////////////////
+TEST_F(JointStatePublisherTest, LimitedPublisher)
+{
+  // Start server
+  ServerConfig serverConfig;
+  serverConfig.SetSdfFile(std::string(PROJECT_SOURCE_PATH) +
+      "/test/worlds/diff_drive_limited_joint_pub.sdf");
+
+  Server server(serverConfig);
+  EXPECT_FALSE(server.Running());
+  EXPECT_FALSE(*server.Running(0));
+
+  server.SetUpdatePeriod(0ns);
+
+  // Check that only the left and right wheel joints are published.
+  std::function<void(const msgs::Model &)> jointStateCb =
+    [&](const msgs::Model &_msg)
+    {
+      bool foundLeftWheelJoint{false},
+           foundRightWheelJoint{false},
+           extra{false};
+
+      for (int i = 0; i < _msg.joint_size(); ++i)
+      {
+        if (_msg.joint(i).name() == "left_wheel_joint")
+          foundLeftWheelJoint = true;
+        else if (_msg.joint(i).name() == "right_wheel_joint")
+          foundRightWheelJoint = true;
+        else
+          extra = true;
+      }
+      EXPECT_TRUE(foundLeftWheelJoint);
+      EXPECT_TRUE(foundRightWheelJoint);
+      EXPECT_FALSE(extra);
+    };
+
+  transport::Node node;
+  node.Subscribe("/world/diff_drive/model/vehicle/joint_state", jointStateCb);
+
+  server.Run(true, 10, false);
```

Comment:
Same here about checking if the callback is called.

Reply:
Added a check. 36b7b3e5f83179be42e8b67d9cd2ecb71dd81046